core/domain/html_validation_service.py

```diff
@@ -718,15 +713,9 @@ def _validate_customization_args_in_tag(tag):
         Error message if the attributes of tag are invalid.
     """
 
-    COMPONENT_TYPES_TO_COMPONENT_CLASSES = {
-        'oppia-noninteractive-collapsible': components.Collapsible,
-        'oppia-noninteractive-image': components.Image,
-        'oppia-noninteractive-link': components.Link,
-        'oppia-noninteractive-math': components.Math,
-        'oppia-noninteractive-tabs': components.Tabs,
-        'oppia-noninteractive-video': components.Video
-    }
-    SIMPLE_COMPONENTS = {
+    COMPONENT_TYPES_TO_COMPONENT_CLASSES = (
+        component_registry.Registry.get_component_type_to_component_classes())
+    SIMPLE_COMPONENTS_TAG_NAMES = {
```

Comment:
Also this should come from the registry. As a general rule, none of these general functions (or the registry itself) should be specifying individual components by name. So make sure your code is clean from that perspective.

Instead, any per-component differences should be specified in the rte-component-definitions JS file (for stuff needed in frontend) or as a property within the RTE class (for stuff needed only in backend). See how interactions work for an example of this. Also add a backend test to make sure that the rte-component-definitions file is consistent with the classes that exist in the backend (same keys etc.) and that all rte-component-definitions have the necessary properties defined (if there isn't a test for this already, though I think there should be).

That will keep the system that you're building solid and maintainable.

Reply:
I checked for component-definition test and I found out that we have a rte component registry defined already in `core/domain/rte_component_registry.py`. The test are also defined to check component definitions. I added the extra functions and test in both these files and removed the new files for component registry. 

But this created a problem of circular dependency [here](https://github.com/oppia/oppia/pull/5362/commits/5779caf1783738286e16f6cec57835b3f5721074#diff-4694a844f6bb2a2a3044c37acfb45d70R72).

One of the way to fix this would be to add these functions in `extensions/rich_text_components/component_utils.py`. Should I do that?

Reply:
I think the fix for this is quite simple, actually. Just remove "from extensions.rich_text_components import components". That should work?

As you can see the interaction_registry.py file doesn't need an import reference to extensions either.

Reply:
I fixed the issue of circular dependency by using pkgutil to load the components module but I get a strange error when I run the `rte_component_registry_test`. 

The error is due to the `test_get_component_types_to_component_classes` method in the test. In this method, I test that the component classes in the types to classes mapping and and the component classes defined in `components.py` are same. I get the error below and from the message both sets are same but still they are not considered equal. I am not able to understand the reason behind this. Can you please help me here. 

```
AssertionError: Items in the first set but not the second:
<class 'components.Collapsible'>
<class 'components.Link'>
<class 'components.Tabs'>
<class 'components.Video'>
<class 'components.Math'>
<class 'components.Image'>
Items in the second set but not the first:
<class 'components.Link'>
<class 'components.Math'>
<class 'components.Tabs'>
<class 'components.Video'>
<class 'components.Collapsible'>
<class 'components.Image'>
```

Reply:
Test the class names, or some primitive value in the class that defines it. Comparing objects can be tricky because they might not be the same instance of the object. You'll probably find that some internal property (like a memory address) differs.

Reply:
Ok, Got it! Thanks!

I have now changed the test function to compare class names and it works fine now, PTAL!